src/device_imxrt.c

```diff
@@ -0,0 +1,278 @@
+  int (*xfer)(uint32_t instance, char *xfer);
+  int (*update_lut)(uint32_t instance, uint32_t seqIndex,
+                         const uint32_t *lutBase, uint32_t seqNumber);
+  int (*get_config)(uint32_t instance, flexspi_nor_config_t *config,
+                         uint32_t *option);
+} flexspi_nor_driver_interface_t;
+#endif
+
+typedef struct {
+  const uint32_t version;  // Bootloader version number
+  const char *copyright;   // Bootloader Copyright
+  void (*runBootloader)(
+      void *arg);             // Function to start the bootloader executing
+  const uint32_t *reserved0;  // Reserved
+  const flexspi_nor_driver_interface_t
+      *flexSpiNorDriver;         // FlexSPI NOR Flash API
+  const uint32_t *reserved1[2];  // Reserved
+} bootloader_api_entry_t;
+
+#define bootloader (*(bootloader_api_entry_t **) (0x0020001c))
+#define flexspi_nor bootloader->flexSpiNorDriver
+
+
+static bool s_flash_irq_disabled;
+
+MG_IRAM void *mg_flash_start(void) {
+  return (void *) 0x0;
+}
+MG_IRAM size_t mg_flash_size(void) {
+  return 8 * 1024 * 1024;
+}
+MG_IRAM size_t mg_flash_sector_size(void) {
+  return 4 * 1024;  // 4k
+}
+MG_IRAM size_t mg_flash_write_align(void) {
+  return 256;
+}
+MG_IRAM int mg_flash_bank(void) {
+  return 0;
+}
+
+MG_IRAM static bool flash_page_start(volatile uint32_t *dst) {
+  char *base = (char *) mg_flash_start(), *end = base + mg_flash_size();
+  volatile char *p = (char *) dst;
+  return p >= base && p < end && ((p - base) % mg_flash_sector_size()) == 0;
+}
+
+// Note: the get_config function below works both for RT1020 and 1060
+#if MG_DEVICE == MG_DEVICE_RT1020
+MG_IRAM static int flexspi_nor_get_config(flexspi_nor_config_t *config) {
+  flexspi_nor_config_t default_config = {
+      .memConfig = {.tag = FLEXSPI_CFG_BLK_TAG,
+                    .version = FLEXSPI_CFG_BLK_VERSION,
+                    .readSampleClkSrc = 1,  // ReadSampleClk_LoopbackFromDqsPad
+                    .csHoldTime = 3,
+                    .csSetupTime = 3,
+                    .controllerMiscOption = MG_BIT(4),
+                    .deviceType = 1,  // serial NOR
+                    .sflashPadType = 4,
+                    .serialClkFreq = 7,  // 133MHz
+                    .sflashA1Size = 8 * 1024 * 1024,
+                    .lookupTable = __FLEXSPI_QSPI_LUT},
+      .pageSize = 256,
+      .sectorSize = 4 * 1024,
+      .ipcmdSerialClkFreq = 1,
+      .blockSize = 64 * 1024,
+      .isUniformBlockSize = false};
+
+  *config = default_config;
+  return 0;
+}
+#else
+MG_IRAM static int flexspi_nor_get_config(flexspi_nor_config_t *config) {
+  uint32_t options[] = {0xc0000000, 0x00};
+  if (!s_flash_irq_disabled) {
+    asm("cpsid i");
+  }
+  uint32_t status =
+      flexspi_nor->get_config(FLEXSPI_NOR_INSTANCE, config, options);
+  if (!s_flash_irq_disabled) {
+    asm("cpsie i");
+  }
+  if (status) {
+    MG_ERROR(("Failed to extract flash configuration: status %u", status));
+  }
+  return status;
+}
```

Comment:
too many checks. It does no harm to disable already disabled IRQs, if you need them to be disabled just do it. These checks only make sense when for some reason you need to keep track of the IRQ state before entering your code and restore it at exit. I guess you do this because sometimes you don't want IRQs to be reenabled.

This has to be written in a compiler agnostic way... plus... we already have a macro for these:
```
		#define MG_ARM_DISABLE_IRQ() asm volatile("cpsid i" : : : "memory")
		#define MG_ARM_ENABLE_IRQ() asm volatile("cpsie i" : : : "memory")
```
These might not be compiler agnostic, are they ?

We also have these:
```
	#if defined(__CC_ARM)
		#define MG_DSB() __dsb(0xf)
	#elif defined(__ARMCC_VERSION)
		#define MG_DSB() __builtin_arm_dsb(0xf)
	#elif defined(__GNUC__) && defined(__arm__) && defined(__thumb__)
		#define MG_DSB() asm("DSB 0xf")
	#elif defined(__ICCARM__)
		#define MG_DSB() __iar_builtin_DSB()
	#else
		#define MG_DSB()
	#endif
```
and behold our own CMSIS_compiler, a reduced one

(And this calls for a revision on what we are doing in other modules...)